dlib/unordered_pair.h

```diff
@@ -65,6 +65,14 @@ namespace dlib
                 - #first and #second contain copies of the items a and b.
         !*/ {}
 
+        unordered_pair (
+            const unordered_pair& p
+        ) = default;
+        /*!
+            ensures
+                - #*this is a copy of p
+        !*/
```

Comment:
The assignment operator does not look trivial, so guessing the rule of three (or actually The Law of The Big Two?) is appropriate here. (?)

Reply:
I mean, this code really is fine as is.  But if visual studio is giving you a warning about there not being like `~unordered_pair() = default;` it's fine to add it.  This is not a resource management deal though so rule of three doesn't apply.

Reply:
No, Visual Studio isn't complaining here. And agree that The Law of The Big Two is what we should indeed talk about.

Reply:
And the `unordered_pair (const unordered_pair& p) = default` is fine too if that's what makes VC happy.  Wasn't sure if you were saying VC was complaining about wanting a defaulted destructor too.

Reply:
The first comment just wanted to say that the assignment operator does not look trivial, and therefore I'm proposing to add a copy constructor and not remove the assignment operator, as in `crc32` and others.

However, it wasn't Visual Studio but gcc who complained about there being an assignment operator but not a copy constructor (as per The Law of The Big Two, I believe):
```
error: implicitly-declared ‘constexpr dlib::unordered_pair<double>::unordered_pair(const dlib::unordered_pair<double>&)’ is deprecated [-Werror=deprecated-copy]
```